modules/Emails/Email.php

```diff
@@ -4235,21 +4238,159 @@ protected function handleBodyInPlainTextFormat($mail)
     }
 
+    /**
+     * @param string $emailField
+     * @return string
+     */
+    public function getEmailAddressConfirmOptInTick($emailField)
+    {
+        global $sugar_config;
+        global $app_list_strings;
+        global $app_strings;
+        global $mod_strings;
+
+        $tickHtml = '';
+
+        if ($sugar_config['email_enable_confirm_opt_in']) {
+            $template = new Sugar_Smarty();
+            $template->assign('APP', $app_strings);
+            $template->assign('APP_LIST_STRINGS', $app_list_strings);
+            $template->assign('MOD', $mod_strings);
+            $template->assign('OPT_IN', $this->getEmailAddressOptInStatus($emailField));
+            $tickHtml = $template->fetch('include/SugarObjects/templates/basic/tpls/displayEmailAddressOptInField.tpl');
+        }
+
+        return $tickHtml;
+    }
+
     /**
      *
+     * @global array $sugar_config
+     * @global \LoggerManager $log
      * @param string $emailField
+     * @return \EmailAddress
+     * @throws RuntimeException
      * @throws InvalidArgumentException
      */
+    public function getEmailAddressConfirmOptIn($emailField)
+    {
+        global $sugar_config;
+
+        if (!$sugar_config['email_enable_confirm_opt_in']) {
+            global $log;
+            $log->warn('Confirm Opt In is not enabled.');
+
+            return false;
+        }
+
+        $emailAddressId = $this->getEmailAddressId($emailField);
+
+        return BeanFactory::getBean('EmailAddresses', $emailAddressId);
+    }
+
+    /**
+     * @param string $id
+     */
+    private function checkOptInFromEmailAddressId($id = '')
+    {
+        global $sugar_config;
+        global $log;
+
+        if ($id === '') {
+            $log->fatal('Empty Email Id');
+        }
+
+        $emailAddresses = BeanFactory::getBean('EmailAddresses');
+        if ($sugar_config['email_enable_auto_send_opt_in']) {
+            /** @var \EmailAddress $emailAddress */
+            $emailAddress = $emailAddresses->retrieve($id);
+            if (
+                ($emailAddress->confirm_opt_in != '1' && empty($emailAddress->opt_in_email_created))
+                || ($_REQUEST['send_opt_in_checkbox'] == 'true')
+            ) {
+                $this->sendOptInEmail($emailAddress);
+            }
+        }
+    }
+
+    /**
+     * @param EmailAddress $emailAddress
+     * @return bool
+     */
+    private function sendOptInEmail(EmailAddress $emailAddress)
+    {
+        global $sugar_config;
+        global $timedate;
+        global $log;
+        global $db;
+
+        require_once __DIR__ . '/../AOW_Actions/actions/actionSendEmail.php';
+
+        if (!$sugar_config['email_enable_confirm_opt_in']) {
+            $log->warning('Confirm Opt In is not enabled.');
+
+            return false;
+        }
+
+
+        if (!$sugar_config['aop']['confirmed_opt_in_template_id']) {
+            $log->fatal('Opt In Email Template is not configured. Please set up in email settings');
```

Comment:
I think it's a "user error", can you show a message to let to know the user have to do some settings here. You can use `SugarApplication::appendErrorMessage()` for e.g.